### PR TITLE
Redispatch NJT rms_norm through dense values for fused kernel

### LINKED ISSUE #191739: RMSNorm on NJT doesn't dispatch to fused implementation
### 🐛 Describe the bug

`F.rms_norm` on a jagged NestedTensor falls back to a decomposition into elementwise ops rather than reaching `_fused_rms_norm`, even when normalizing purely over trailing dense dims (the ragged dim is untouched). The dense path uses the fused kernel.

### Repro

```python
import time
import torch
import torch.nn.functional as F
from torch.utils._python_dispatch import TorchDispatchMode


class OpLog(TorchDispatchMode):
    def __init__(self):
        self.ops = []

    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        self.ops.append(func._overloadpacket.__name__)
        return func(*args, **(kwargs or {}))


dim, n = 512, 52424
values = torch.randn(n, dim, device="cuda")
lengths = [149] * (n // 149) + [n - 149 * (n // 149)]
offsets = torch.tensor([0, *torch.tensor(lengths).cumsum(0).tolist()], device="cuda")
njt = torch.nested.nested_tensor_from_jagged(
    values, offsets=offsets, min_seqlen=0, max_seqlen=149
)
weight = torch.ones(dim, device="cuda")

for name, x in (("dense", values), ("njt", njt)):
    with OpLog() as log:
        F.rms_norm(x, (dim,), weight=weight, eps=1e-6)
    print(name, sorted(set(log.ops)))
```

### Related issues and PRs

- #135872 — *Support rms_norm() for NJT*, landed 2024-09-17 (`a8382847f452`). This is the PR that added the `__torch_function__` block quoted above. Its own description says it "SymInt-ifies `rms_norm()`, allowing NJT to use the same decomp", so routing through the decomposition was the original design, not a regression.
- #153666 — *Fused RMSNorm implementation*, added `_fused_rms_norm`; landed 2025-07-22 after two reverts, followed by #159317 (*Fused RMSNorm Housekeeping*, 2025-07-29). It post-dates #135872 by ~10 months, and `_fused_rms_norm` still does not appear anywhere under `torch/nested/`.
- #164289 — *Remove CompositeImplicitAutograd from `_fused_rms_norm`*, closed 2025-10-10 without landing. It reports the same class of problem from the dense/`aot_eager` side: "The failure of bitwise equivalence is because aot_eager decomposes rms_norm." `_fused_rms_norm` still carries a `CompositeImplicitAutograd[alias]` registration in `2.14.0.dev20260707` (checked with `torch._C._dispatch_dump("aten::_fused_rms_norm")`).
- #141916 — *rms_norm forward performance analysis* (open), dense-side `rms_norm` performance.

### Versions

PyTorch version: 2.14.0.dev20260707+cu130
CUDA used to build PyTorch: 13.0
Python version: 3.12.13 (main, Jun 23 2026, 15:18:55) [Clang 22.1.3] (64-bit runtime)
Python platform: Linux-6.11.0-1016-nvidia-x86_64-with-glibc2.39
OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Libc version: glibc-2.39

Is CUDA available: True
GPU models and configuration: GPU 0: NVIDIA H200 (143771 MiB)
Nvidia driver version: 570.195.03
CUDA runtime version: 12.9.86
cuDNN version: 9.10.1

CPU: Intel(R) Xeon(R) Platinum 8468, 128 threads, 2 sockets

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ

### PR DESCRIPTION
**Pull request by UlviDemirsoy**
## Issue

Fixes #191739

## Summary

`F.rms_norm` on jagged NestedTensor was falling through the elementwise decomp even when normalizing only over trailing dense dims. After the existing ragged-dim validation, redispatch through `input.values()` into dense `torch.rms_norm` (so CUDA can hit `_fused_rms_norm`), then rewrap with nested views (`nested_view_from_values_like`) to keep NestedTensor autograd intact.

## Checklist

- [ ] Passes lint (`spin fixlint`) — full `spin`/`lintrunner` init is unsupported on Windows here; `ruff check` on the touched files passed locally. Will rely on CI lint.
- [x] Added/updated tests
- [ ] Updated documentation (if applicable)
- [x] Included benchmark / dispatch results (for PRs impacting perf)

Local checks (`torch 2.11.0+cu128`, RTX 3060):
- Dispatch: dense and NJT both hit `_fused_rms_norm` (no `mean`/`rsqrt` decomp on CUDA)
- Profiler: NJT calls land in `vectorized_layer_norm_kernel`
- `python test/test_nestedtensor.py -k rms_norm -v` → OK
- Contiguous OpInfo forward/backward covered; noncontig-holes backward skipped because the values buffer includes hole contributions unlike the unbind reference
- Correctness: NJT vs dense `max|diff| = 0` for forward and `dvalues`/`dweight`

### Perf / memory (local microbench)

Same jagged layout (~17–20k tokens × `dim=4096`, batch 64, random lengths 64–512), CUDA:

| Path | Forward | Fwd+Bwd | Peak mem delta (fwd / fb) |
|------|---------|---------|---------------------------|
| Dense fused | 7.79 ms | 29.4 ms | 540 / 1080 MB |
| **NJT fused (this PR)** | **7.84 ms** | **30.2 ms** | **540 / 1080 MB** |
| Old composite decomp | 17.2 ms | 70.1 ms | 810 / 1890 MB |

- ~2.2× faster fwd / ~2.3× faster fwd+bwd vs the previous NJT decomp path
- Negligible overhead vs dense fused (~50 µs; same peak memory)
- Old path allocated extra intermediates (`pow`/`mean`/`rsqrt`); fused path does not

## BC-breaking?

No

## AI assistance

Prepared with Cursor agent assistance. I reviewed the NJT `rms_norm` path, kept the existing ragged-dim guard, matched the SDPA-style nested view rewrap for autograd, and verified fused dispatch, correctness, perf/memory, and OpInfo/`-k rms_norm` locally.

### CODE REVIEW

**Review by Mr-Ashish — APPROVED** [review]
## 🏴‍☠️ Torii — APPROVE
**Score:** 89/100 · **Confidence:** high

Torii: APPROVE

_Full review is the PR comment with marker `<!-- torii-review pr=191851`._

**Comment by soulitzer** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
